### ISSUE TITLE
Cannot find step definition used in scenario outline

### ISSUE THREAD

**Issue by whazzmaster** [alexkrechik/VSCucumberAutoComplete#194]
![outline](https://user-images.githubusercontent.com/102357/41546031-2a20a3c8-72e2-11e8-854b-e6b653391aff.png)

I have an existing step definition, but when used in a scenario outline with `<runName>` instead of `"run1"` I get a warning:

```
[cucumberautocomplete] Was unable to find step for "And I rename the run to <runName>"
```

The step is defined thusly:

```typescript
When(/^I rename the run to "([^"]*)"$/, (name) =>
  // ... whatever
);
```

I feel like this may be a configuration issue but it's unclear to me how to solve it, as it appears that scenario outline steps using angle brackets _should_ be supported in the current version.

Currently using VSCode 1.24.1 and VSCucumberAutoComplete 2.11.0.

**Comment by stephanmullerNL** [comment]
I'm encountering the same issue. The problem seems to go away when I remove the double quotes from the regex, like this:

```
When(/^I rename the run to (.+)$/, (name) =>
  // ... whatever
);
```


I prefer using quotes in my scenarios because it makes it a lot easier to see which parts of a step are variable when it's _not_ an outline, but I guess in outlines they could be left out.  That does mean that a step can't be re-used between regular scenarios and scenario outlines without losing some part of readabilty.

A fix would be greatly appreciated :)

**Comment by alexkrechik** [comment]
Unfortunately, for the most cases step definition regex should match gherkin line string, but, in this case, it doesn't.
Usually such error could be fixed with `cucumberautocomplete.customParameters`, but, in this case. I'm not sure, that `"([^"]*)"` could be changes with something. So, should review gherkin files parsing to fix this in some way.

**Comment by cameroncooke** [comment]
If your step is:

```gherkin
And I rename the run  to <runName>
```

Then your step definition regular expression is wrong:

```ruby
When(/^I rename the run to "([^"]*)"$/, (name) =>
  // ... whatever
);
```

as `"([^"]*)"` matches anything but a `"`. This means that to satisify the step the expression would need to be:

```ruby
When(/^I rename the run to (.+)$/, (name) =>
  // ... whatever
);
```

Alternatively change your step to:

```gherkin
And I rename the run  to "<runName>"
```

Which would work with the current step definition.

Try for yourself:

Original step def (doesn't match):
https://regex101.com/r/fheBcs/1

Fixed step def (matches):
https://regex101.com/r/lA4wMn/1

**Comment by whazzmaster** [comment]
@cameroncooke This issue is in the context of a [scenario outline](https://relishapp.com/cucumber/cucumber/docs/gherkin/scenario-outlines), not a regular scenario. In that context, you must use bare angle brackets in order for cucumber to substitute the tabular data defined later. See the link above re: **placeholders** in scenario outlines.

**Comment by cameroncooke** [comment]
Now I've got the opposite problem, my scenario outline has a step like:

```gherkin
When I change the quantity of the basket product to <Quantity>
```

Example with data populated:

```gherkin
When I change the quantity of the basket product to 5
```

with step definition:

```regex
^I change the quantity of the basket product to ([0-9]*)$
```

I get the error:

```
[cucumberautocomplete] Was unable to find step for "When I change the quantity of the basket product to <Quantity>"
```

I assume this is because VSCucumberAutoComplete doesn't know what the replacement type is and the regex in the step def quite specifically says it's a number.

Full example:

```gherkin
  Scenario Outline: Do something
    Given I change the quantity of the basket product to <Quantity>
    When bla
    Then bla
    Examples:
      | Quantity |
      | 2        |
      | 5        |
```

Changing the step def to:
```
I change the quantity of the basket product to (.+)
``` 

Fixes the issue, I assume this is a limitation of VSCucumberAutoComplete, it could read the values from the outline table and use that to substitute the replacement tag when attempting to find matching step definitions?

**Comment by whazzmaster** [comment]
Yep, the gherkin code works correctly when executed but it's a tough parsing problem from the perspective of the extension.

**Comment by cameroncooke** [comment]
> @whazzmaster This issue is in the context of a scenario outline, not a regular scenario. In that context, you must use bare angle brackets in order for cucumber to substitute the tabular data defined later. See the link above re: placeholders in scenario outlines.

Yes I know but the <placeholders> and substituted at run-time for values in the tabular data, then cucumber does it's usual regular expression matching of the step definition. 

In your original example, the regular expression you're using for the step definition is looking for a string value sub-match surrounded by quotes.

So unless the step looks like:

```
  Scenario Outline: Do something
    Given And I rename the run to <runName>
    When bla
    Then bla
    Examples:
      | runName            |
      | "some vaue"        |
      | "another value"    |
```

It's never got to match the step def regular expression because of the requirement of matching the quotes.

**Comment by cameroncooke** [comment]
`VSCucumberAutoComplete` does its regex matching at design-time so only sees `Given And I rename the run to <runName>` when looking for matches, while cucumber at run-time sees `Given And I rename the run to somerunname` when looking for matches. So the regex will never satisfy both scenarios unless made more generic.

**Comment by whazzmaster** [comment]
I misunderstood your original comment; apologies. Yes, I agree with you-- the code I'm working with is trying to use the same step definition in both scenarios and scenario outlines but due to the presence or absence of double-quotes in the context the extension can't support both usages.

**Comment by cameroncooke** [comment]
I see so you're trying to use a non-outline specific definition as part of scenario outline with replacements?

So from a normal scenario, you're using it like so:

```gherkin
Given And I rename the run to "somerunname"
```

and from an outline 

```gherkin
Given And I rename the run to <runName>
    Examples:
      | runName            |
      | "some vaue"        |
      | "another value"    |
```

I can't see why that isn't possible, can't you just change the outline one to:

```gherkin
Given And I rename the run to "<runName>"
    Examples:
      | runName            |
      | some vaue        |
      | another value    |
```

**Comment by ilyapuchka** [comment]
I faced the same problem with step defined as `"I use the example name (?:Alice|Bob)"`, where it can actually autocomplete while I'm typing `"Given I use the example name"` in a feature file, but fails to find the step (and so no jump-to also) when I change `(?:Alice|Bob)` pattern to `<name>` that references value from example table.
Changing to `(.+)` makes autocompletion work but requires to assert values in steps instead of having an error that step is undefined.
